### PR TITLE
Fix/20260 layers property setter

### LINKED ISSUE #20260: Model weights not saved if a custom layer class contains a list of layers named self._layers
Hi everyone,

I would like to point out a problem that I found while developing a custom model: layer's weights are not saved if a Model subclass initializes  a custom layer that contains sublayers in a class-level list variable called `self._layers`. The code below proves the issue and should illustrates it better.

```python
import os
from pathlib import Path

import keras
import numpy as np


@keras.saving.register_keras_serializable(package="KerasTest", name="CustomLayer")
class CustomLayer(keras.Layer):

    def __init__(self, bugged: bool = False, name="test_layer", **kwargs):
        super(CustomLayer, self).__init__(name=name, **kwargs)
        self._bugged = bugged
        if self._bugged:
            self._layers = [
                keras.layers.Dense(64, activation='relu'),
                keras.layers.Dense(32, activation='relu'),
                keras.layers.Dense(10, activation='softmax')
            ]
        else:
            self._custom_layers = [
                keras.layers.Dense(64, activation='relu'),
                keras.layers.Dense(32, activation='relu'),
                keras.layers.Dense(10, activation='softmax')
            ]

    def call(self, inputs):
        x = inputs
        layer_list = self._layers if self._bugged else self._custom_layers
        for layer in layer_list:
            x = layer(x)
        return x

    def get_config(self):
        base_config = super().get_config()
        config = {
            "bugged": self._bugged
        }
        return {**base_config, **config}


@keras.saving.register_keras_serializable(package="KerasTest", name="TestModel")
class TestModel(keras.Model):

    def __init__(self, bugged: bool = False, name="test_model", **kwargs):
        super(TestModel, self).__init__(name=name, **kwargs)
        self._bugged = bugged
        self._custom_layer = CustomLayer(bugged=bugged)

    def call(self, inputs):
        return self._custom_layer(inputs)

    def get_config(self):
        base_config = super().get_config()
        config = {
            "bugged": self._bugged
        }
        return {**base_config, **config}


def test_model(bugged: bool = False):
    output_path = Path("./output")
    model_name_prefix = "bugged" if bugged else "fixed"

    # Dataset generation
    num_samples = 1000
    input_data = keras.random.uniform(shape=(num_samples, 32))
    labels = keras.random.randint(minval=0, maxval=10, shape=(num_samples,))
    labels = keras.utils.to_categorical(labels, num_classes=10)

    # Test bugged model
    model = TestModel(bugged=bugged)
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    model.fit(input_data, labels, epochs=5, batch_size=32, validation_split=0.2)
    trained_weights = model.get_weights()

    # Save and load model
    output_path.mkdir(parents=True, exist_ok=True)
    model.save(output_path / f"{model_name_prefix}_model.keras")
    loaded_model = keras.saving.load_model(output_path / f"{model_name_prefix}_model.keras")
    loaded_weights = loaded_model.get_weights()

    print(f"------------ {model_name_prefix.capitalize()} - Compare trained weights to loaded weights -------------")
    for i, (uw, tw) in enumerate(zip(trained_weights, loaded_weights)):
        comparison = np.array_equal(uw, tw)
        print(f"Layer {i} -> Weights match: {comparison}")


if __name__ == '__main__':
    os.environ["KERAS_BACKEND"] = "tensorflow"
    test_model(bugged=True)
    test_model(bugged=False)
```

The output is:

```
------------ Bugged - Compare trained weights to loaded weights -------------
Layer 0 -> Weights match: False
Layer 1 -> Weights match: False
Layer 2 -> Weights match: False
Layer 3 -> Weights match: False
Layer 4 -> Weights match: False
Layer 5 -> Weights match: False
------------ Fixed - Compare trained weights to loaded weights -------------
Layer 0 -> Weights match: True
Layer 1 -> Weights match: True
Layer 2 -> Weights match: True
Layer 3 -> Weights match: True
Layer 4 -> Weights match: True
Layer 5 -> Weights match: True
```

I think that the same problem has been solved for the `Model` class by declaring a setter method (line 170). Perhaps it is possible to use the same approach for the `Layer` class.

### PR DESCRIPTION
**Pull request by pctablet505**
Fixes #20260

### **Problem**
Keras internally uses tracking attributes (e.g. `_layers`, `_metrics`, `_trainable_variables`) inside the `Layer` class. If a user subclass explicitly assigns to them (`self._layers = [...]`), the internal tracker's list is silently wiped. This leads to subtle bugs—such as model weights not saving properly—and crucially, gives no warning or indication of failure to the developer.

### **Solution**
This PR adds a safety mechanism in `Layer.__setattr__` that explicitly warns users if they overwrite any Keras tracking attributes.

#### **Key Improvements:**
* **Reserved Attributes Tracking:** Added `_RESERVED_LAYER_ATTRIBUTES` and an exempt module list `_RESERVED_ATTR_EXEMPT_MODULES` to specify valid internal callers (e.g. `Sequential` or `Functional`).
* **Dynamic Warning System**: When an attribute is overridden, the check dynamically inspects the caller module via `inspect.currentframe().f_back`. This correctly warns offending user assignments while allowing valid inherited/internal configurations (like `Sequential.__init__`) to bypass safety without triggering false positive warnings. 
* **Tests Included**: Validates `BadLayer` overriding `_layers` emits a structured `UserWarning`, and `Sequential` initializes cleanly.

### **Testing:**
- Verified by adding `test_reserved_attribute_warning` to `layer_test.py`.
- Checked against `saving_lib_test.py` ensuring that standard implementations of `Sequential` no longer inadvertently emit alerts during saves mapping.

### CODE REVIEW

**Review by hertschuh — CHANGES_REQUESTED** [review]
This seems very ad-hoc and brittle. We should find a better way to fix this.

I think a simpler and cleaner criteria that doesn't involve hardcoded _RESERVED_LAYER_ATTRIBUTES and hardcoded _RESERVED_ATTR_EXEMPT_MODULES is to check whether what you're reassigning is one of the lists that the tracker itself uses (in it's config). Maybe we can cache the attribute name in the config too to be able to check quickly. And we'll need to pass the attribute name to `_tracker.track()`, which we'll need anyway to fix https://github.com/keras-team/keras/issues/18601

File: keras/src/utils/tracking.py [hunk 1]
```diff
@@ -67,12 +67,16 @@ def __init__(self, config, exclusions=None):
         self.locked = False
         self._lock_violation_msg = None
         self.exclusions = exclusions or {}
+        self.attr_to_store_name = {}
```
Comment (hertschuh) [inline keras/src/utils/tracking.py]: Let's call it something like `self.tracking_collections_attr_names` and make it a `set()`. We won't need the values per my other comment.

File: keras/src/utils/tracking.py [hunk 2]
```diff
@@ -67,12 +67,16 @@ def __init__(self, config, exclusions=None):
...
+        for k, v in config.items():
+            if len(v) > 2:
+                self.attr_to_store_name[v[2]] = k
```
Comment (hertschuh) [inline keras/src/utils/tracking.py]: No need to support backwards compatibility, remove the `if len(v) > 2`.

File: keras/src/utils/tracking.py [hunk 3]
```diff
@@ -67,12 +67,16 @@ def __init__(self, config, exclusions=None):
...
 
-    def track(self, attr):
+    def track(self, attr, name=None):
```
Comment (hertschuh) [inline keras/src/utils/tracking.py]: Let's remove the `name` attribute, it's not used. We'll add it in a separate PR when neeeded.

File: keras/src/layers/layer.py [hunk 4]
```diff
@@ -1576,6 +1584,23 @@ def __str__(self):
         return self.__repr__()
 
     def __setattr__(self, name, value):
+        # Warn if user code reassigns a reserved tracked attribute.
+        if (
+            hasattr(self, "_tracker")
+            and name in self._tracker.attr_to_store_name
+            and tracking.is_tracking_enabled()
+        ):
```
Comment (hertschuh) [inline keras/src/layers/layer.py]: Let's remove the `tracking.is_tracking_enabled()` part of the test. You should never re-assign a tracking collection, even if you're not tracking temporarily.

File: keras/src/layers/layer.py [hunk 5]
```diff
@@ -1576,6 +1584,23 @@ def __str__(self):
...
+            store_name = self._tracker.attr_to_store_name[name]
+            store_list = self._tracker.config[store_name][1]
+            if id(value) != id(store_list):
```
Comment (hertschuh) [inline keras/src/layers/layer.py]: Let's remove this test `if id(value) != id(store_list):`.

You should never reassign a tracking collection.

File: keras/src/layers/layer.py [hunk 6]
```diff
@@ -1576,6 +1584,23 @@ def __str__(self):
...
+                warnings.warn(
+                    f"`{name}` is a reserved attribute in Keras layers and "
+                    "should not be used as a variable name in a Layer "
+                    "subclass. Assigning to it can break weight saving, "
+                    "metric tracking, and other functionality. "
+                    f"Please use a different attribute name.",
+                    stacklevel=2,
+                )
```
Comment (hertschuh) [inline keras/src/layers/layer.py]: I think we could make this `raise AttributeError(...)` instead. I don't see a valid use case for doing this.

It might break some people, but it would be shocking of this didn't break in other ways before.

File: keras/src/layers/layer.py [hunk 7]
```diff
@@ -354,21 +354,29 @@ def _initialize_tracker(self):
...
                 "non_trainable_variables": (
                     lambda x: isinstance(x, backend.Variable)
                     and not x.trainable,
                     non_trainable_variables,
+                    "_non_trainable_variables",
+                ),
+                "metrics": (
+                    lambda x: isinstance(x, Metric),
+                    metrics,
+                    "_metrics",
```
Comment (hertschuh) [inline keras/src/layers/layer.py:369]: Please do the same for all the other Trackers in the whole codebase (there are only 4 or 5 total). We don't want to maintain backwards compatibility and we want the same behavior everywhere.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks!

Can you also test that this doesn't break KerasHub?

File: keras/src/metrics/metric.py [hunk 8]
```diff
@@ -90,20 +90,31 @@ def result(self):
...
             }
         )
+        # Now safe to assign other attributes
+        self.name = name or auto_name(self.__class__.__name__)
+        self._dtype_policy = dtype_policies.get(dtype or backend.floatx())
+        self._dtype = self._dtype_policy.compute_dtype
+        # Use object.__setattr__ to bypass tracking wrap so self._metrics and
+        # self._variables stay as the same list objects the Tracker references.
+        object.__setattr__(self, "_metrics", metrics)
+        object.__setattr__(self, "_variables", variables)
```
Comment (hertschuh) [inline keras/src/metrics/metric.py:117]: You should just do what was done before, i.e. do this before creating the tracker.
        self._metrics = []
        self._variables = []
Did that not work?

It should work because everything is protected by `hasattr(self, "_tracker")`

I only commented in this file, but please apply this to the other files too.

File: keras/src/metrics/metric.py [hunk 9]
```diff
@@ -235,6 +246,18 @@ def from_config(cls, config):
...
+            hasattr(self, "_tracker")
+            and name in self._tracker.tracking_collections_attr_names
+            and hasattr(self, name)
+        ):
+            raise AttributeError(
+                f"`{name}` is a reserved attribute in Keras metrics and "
+                "should not be used as a variable name in a Metric "
+                "subclass. Assigning to it can break metric state "
+                "tracking. Please use a different attribute name."
+            )
```
Comment (hertschuh) [inline keras/src/metrics/metric.py:260]: Instead move this as a new `if` under the existing `if hasattr(self, "_tracker"):` below.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Can you also look into why the export tests are failing?
